### ISSUE TITLE
`Trainer` loads model weights twice when resuming from checkpoint

### ISSUE THREAD

**Issue by JamesDeAntonis** [huggingface/transformers#13607]
## Environment info

- `transformers` version: 4.10.0
- Platform:
- Python version: 3.8
- PyTorch version (GPU?): 1.8
- Tensorflow version (GPU?):
- Using GPU in script?: yes
- Using distributed or parallel set-up in script?: yes

## Information

Model I am using (Bert, XLNet ...): `T5` (`t5-3b`)

The problem arises when using:
* [x] the official example scripts: (give details below)
* [ ] my own modified scripts: (give details below)

The tasks I am working on is:
* [ ] an official GLUE/SQUaD task: (give the name)
* [ ] my own task or dataset: (give details below)

## To reproduce

1. Try to do training with resumption from checkpoint
2. Observe that model weights are loaded twice

## Expected behavior

The model weights are loaded once [here](https://github.com/huggingface/transformers/blob/v4.10.0/src/transformers/trainer.py#L260) as passed to the constructor, and then again [here](https://github.com/huggingface/transformers/blob/v4.10.0/src/transformers/trainer.py#L1074).  We're using a large model so this causes unnecessary OOM.

My proposal: I think the constructor should simply take the class of the model itself.  No model weights are loaded until `self.train`.  Inside `self.train`, the model weights path, optimizer path, etc are deduced using logic of whether `resume_from_checkpoint` is provided, `model_name_or_path`, etc

**Comment by JamesDeAntonis** [comment]
On a similar note, why do we put the optimizer on cuda [here](https://github.com/huggingface/transformers/blob/v4.10.0/src/transformers/trainer.py#L1621)?  Can we instead keep the optimizer on cpu?  That would save more memory

**Comment by sgugger** [comment]
Your proposal would be a major breaking change as the `Trainer` initialization wouldn't take the model anymore, and the model attribute of the `Trainer` would not be available after the init anymore. We don't do breaking changes between minor releases, and it might even be too big for a major release.

As for your question about the the optimizer, it needs to be on CUDA like the model to be able to perform the optimizer step. I invite you to remove the line and look for yourself the error message of mismatched devices you will get.

If you want to save memory this way, you can have a look at the ZeRO-offload integration with DeepSpeed, which does offload the optimzier stat and the gradients to the CPU.

**Comment by JamesDeAntonis** [comment]
Thanks for the quick reply.  That's good to know; maybe we can live without the change.

>I invite you to remove the line and look for yourself the error message of mismatched devices you will get.

Somehow, my teammate @dblakely is able to remove the line with no error.  Not sure why.
